internal/operator/statefulset/statefulset.go

```diff
@@ -0,0 +1,217 @@
+func (r _resource) GetExistingResourceNames(ctx component.OperatorContext, etcd *druidv1alpha1.Etcd) ([]string, error) {
+	resourceNames := make([]string, 0, 1)
+	objectKey := getObjectKey(etcd)
+	sts, err := r.getExistingStatefulSet(ctx, etcd)
+	if err != nil {
+		return nil, druiderr.WrapError(err,
+			ErrGetStatefulSet,
+			"GetExistingResourceNames",
+			fmt.Sprintf("Error getting StatefulSet: %v for etcd: %v", objectKey, etcd.GetNamespaceName()))
+	}
+	if sts == nil {
+		return resourceNames, nil
+	}
+	if metav1.IsControlledBy(sts, etcd) {
+		resourceNames = append(resourceNames, sts.Name)
+	}
+	return resourceNames, nil
+}
+
+// Sync creates or updates the statefulset for the given Etcd.
+func (r _resource) Sync(ctx component.OperatorContext, etcd *druidv1alpha1.Etcd) error {
+	var (
+		existingSTS *appsv1.StatefulSet
+		err         error
+	)
+	objectKey := getObjectKey(etcd)
+	if existingSTS, err = r.getExistingStatefulSet(ctx, etcd); err != nil {
+		return druiderr.WrapError(err,
+			ErrSyncStatefulSet,
+			"Sync",
+			fmt.Sprintf("Error getting StatefulSet: %v for etcd: %v", objectKey, etcd.GetNamespaceName()))
+	}
+	// There is no StatefulSet present. Create one.
+	if existingSTS == nil {
+		return r.createOrPatch(ctx, etcd)
+	}
+
+	// StatefulSet exists, check if TLS has been enabled for peer communication, if yes then it is currently a multistep
+	// process to ensure that all members are updated and establish peer TLS communication.
+	if err = r.handlePeerTLSChanges(ctx, etcd, existingSTS); err != nil {
+		return druiderr.WrapError(err,
+			ErrSyncStatefulSet,
+			"Sync",
+			fmt.Sprintf("Error while handling peer URL TLS change for StatefulSet: %v, etcd: %v", objectKey, etcd.GetNamespaceName()))
+	}
+	return r.createOrPatch(ctx, etcd)
+}
+
+// TriggerDelete triggers the deletion of the statefulset for the given Etcd.
+func (r _resource) TriggerDelete(ctx component.OperatorContext, etcd *druidv1alpha1.Etcd) error {
+	objectKey := getObjectKey(etcd)
+	ctx.Logger.Info("Triggering deletion of StatefulSet", "objectKey", objectKey)
+	if err := r.client.Delete(ctx, emptyStatefulSet(etcd)); err != nil {
+		if errors.IsNotFound(err) {
+			ctx.Logger.Info("No StatefulSet found, Deletion is a No-Op", "objectKey", objectKey.Name)
+			return nil
+		}
+		return druiderr.WrapError(err,
+			ErrDeleteStatefulSet,
+			"TriggerDelete",
+			fmt.Sprintf("Failed to delete StatefulSet: %v for etcd %v", objectKey, etcd.GetNamespaceName()))
+	}
+	ctx.Logger.Info("deleted", "component", "statefulset", "objectKey", objectKey)
+	return nil
+}
+
+// getExistingStatefulSet gets the existing statefulset if it exists.
+// If it is not found, it simply returns nil. Any other errors are returned as is.
+func (r _resource) getExistingStatefulSet(ctx component.OperatorContext, etcd *druidv1alpha1.Etcd) (*appsv1.StatefulSet, error) {
+	sts := emptyStatefulSet(etcd)
+	if err := r.client.Get(ctx, getObjectKey(etcd), sts); err != nil {
```

Comment:
This could be metadata-only get instead of the full object.

Reply:
done